test/bundling.js

```diff
@@ -0,0 +1,827 @@
+      }
+    });
+    extendElements(task, [1, 2, 3], function(err, resp) {
+      expect(resp.field1).to.deep.equal([1, 2, 3]);
+      callback();
+    });
+    extendElements(task, [4, 5, 6], function(err, resp) {
+      done(new Error('should not reach'));
+    });
+    var cancelId = task._data[task._data.length - 1].emitter._eventId;
+
+    extendElements(task, [7, 8, 9], function(err, resp) {
+      expect(resp.field1).to.deep.equal([7, 8, 9]);
+      callback();
+    });
+
+    task.cancel(cancelId);
+    task.run();
+  });
+});
+
+describe('Executor', function() {
+  function apiCall(request, callback) {
+    callback(null, request);
+  }
+  function failing(request, callback) {
+    callback(new Error('failure'));
+  }
+
+  function newExecutor(options) {
+    return new bundling.BundleExecutor(options, new gax.BundleDescriptor(
+      'field1', ['field2'], 'field1', byteLength));
+  }
+
+  function schedule(executor, func, request, callback) {
+    var emitter = new event_emitter.BundleEventEmitter(executor, callback);
+    executor.schedule(func, request, emitter);
+    return emitter;
+  }
+
+  it('groups api calls by the id', function() {
+    var executor = newExecutor(new gax.BundleOptions({'delayThreshold': 100}));
+    schedule(executor, apiCall, createSimple([1, 2], 'id1'));
+    schedule(executor, apiCall, createSimple([3], 'id2'));
+    schedule(executor, apiCall, createSimple([4, 5], 'id1'));
+    schedule(executor, apiCall, createSimple([6], 'id2'));
+
+    expect(executor._tasks).to.have.property('id1');
+    expect(executor._tasks).to.have.property('id2');
+    expect(_.size(executor._tasks)).to.eq(2);
+
+    var task = executor._tasks.id1;
+    expect(task._data.length).to.eq(2);
+    expect(task._data[0].elements).to.eql([1, 2]);
+    expect(task._data[1].elements).to.eql([4, 5]);
+
+    task = executor._tasks.id2;
+    expect(task._data.length).to.eq(2);
+    expect(task._data[0].elements).to.eql([3]);
+    expect(task._data[1].elements).to.eql([6]);
+
+    for (var bundleId in executor._timers) {
+      clearTimeout(executor._timers[bundleId]);
+    }
+  });
+
+  it('emits errors when the api call fails', function(done) {
+    var executor = newExecutor(new gax.BundleOptions({'delayThreshold': 100}));
+    var callback = sinon.spy(function(err, resp) {
+      expect(err).to.be.an.instanceOf(Error);
+      if (callback.callCount == 2) {
+        done();
+      }
+    });
+    var emitter = schedule(
+        executor, failing, createSimple([1], 'id'), callback);
+    var emitter2 = schedule(
+        executor, failing, createSimple([2], 'id'), callback);
+
+    emitter.runNow();
+  });
+
+  describe('with events', function() {
+    var executor = newExecutor(new gax.BundleOptions({'delayThreshold': 100}));
+    var spyApi;
+
+    function timedAPI(request, callback) {
+      var canceled = false;
+      setTimeout(function() {
+        if (!canceled) {
+          callback(null, request);
+        }
+      }, 10);
```

Comment:
Here, too -- it seems like we're using the real clock in a lot of these tests. Is this the best way to do this? Or can we guarantee execution order without setting an actual timeout?